packages/node-core/src/configure/migration-service/migration.ts

```diff
@@ -18,60 +61,115 @@ import {
 } from '../../utils/sync-helper';
 import {NodeConfig} from '../NodeConfig';
 
+type RemovedIndexes = Record<string, IndexesOptions[]>;
+
+const logger = getLogger('db-manager');
+
 export class Migration {
   private sequelizeModels: ModelStatic<any>[] = [];
-  private rawQueries: string[] = [];
+  private mainQueries: string[] = [];
   private readonly historical: boolean;
+  private extraQueries: string[] = [];
+  private foreignKeyMap: Map<string, Map<string, SmartTags>> = new Map<string, Map<string, SmartTags>>();
+  private useSubscription: boolean;
+  private enumTypeMap: Map<string, string> = new Map<string, string>();
+  private removedIndexes: RemovedIndexes = {};
 
   constructor(
     private sequelize: Sequelize,
     private storeService: StoreService,
     private schemaName: string,
     private config: NodeConfig,
-    private enumTypeMap: Map<string, string>
+    private dbType: SUPPORT_DB
   ) {
     this.historical = !config.disableHistorical;
+    this.useSubscription = config.subscription;
+    if (this.useSubscription && dbType === SUPPORT_DB.cockRoach) {
+      this.useSubscription = false;
+      logger.warn(`Subscription is not support with ${this.dbType}`);
+    }
   }
 
   static async create(
     sequelize: Sequelize,
     storeService: StoreService,
     schemaName: string,
-    graphQLSchema: GraphQLSchema,
+    currentSchema: GraphQLSchema | null,
     config: NodeConfig,
-    logger: Pino.Logger
+    dbType: SUPPORT_DB
   ): Promise<Migration> {
-    const modelsRelationsEnums = getAllEntitiesRelations(graphQLSchema);
-    const enumTypeMap = new Map<string, string>();
-    for (const e of modelsRelationsEnums.enums) {
-      await syncEnums(sequelize, SUPPORT_DB.postgres, e, schemaName, enumTypeMap, logger);
+    const migration = new Migration(sequelize, storeService, schemaName, config, dbType);
+    await migration.init(currentSchema);
+    return migration;
+  }
+  async init(currentSchema: GraphQLSchema | null): Promise<void> {
+    const CurrentModelsRelationsEnums = getAllEntitiesRelations(currentSchema);
+
+    // Should load all keys and enums of currentSchema, as nextSchema will be added from the migration execution
+    this.loadExistingForeignKeys(CurrentModelsRelationsEnums.relations);
+    await this.loadExisitingEnums(CurrentModelsRelationsEnums.enums);
+
+    if (this.useSubscription) {
+      this.extraQueries.push(createSendNotificationTriggerFunction(this.schemaName));
     }
```

Comment:
I think rather than these mutating Migration, they should return values and then be passed in as constructor args. It adds further safety that they are called in the correct place and have no side-effects